### DISCUSSION TITLE
Customize "Avalonia Application" title in MacOS

### DISCUSSION THREAD

**Discussion by alessandrofrancesconi** [AvaloniaUI/Avalonia#9346] [Q&A] [answered]
I've noticed that when running my project after `dotnet publish` on MacOS (Catalina), the title "Avalonia Application" appears in the upper bar

![titlebar](https://user-images.githubusercontent.com/1595223/74111345-55814680-4b94-11ea-99bb-5146a8ef42cb.JPG)

Also there is an "About" submenu that opens this dialog:

![about](https://user-images.githubusercontent.com/1595223/74111348-5d40eb00-4b94-11ea-8dd7-a1789ab7620d.JPG)

How can I change this and maybe replace with custom graphics?

**Comment by kekekeks** [comment, marked as answer]
1) Set `Name` property on the `Application` in your `App.xaml`
2) Add a customized native app menu to your `App.xaml`:
```xml
  <NativeMenu.Menu>
    <NativeMenu>
      <NativeMenuItem Header="About My App" Command="{Binding AboutCommand}" />
    </NativeMenu>
  </NativeMenu.Menu>
```
it will replace the default one.

We probably need that in our default app template.

**Comment by sandermvanvliet** [comment]
I've just found this issue after some googling and wanted to add some additional details.

When you set `Name` in your `App.xaml` but package your app in a bundle you'll find that if you set `CFBundleName` in the `Info.plist` to something else then macOS will use the value from `CFBundleName`

For example:

**App.xaml**
```xml
<Application
   ...
   Name="Test app">
</Application>
```

**Info.plist**
```xml
<plist version="1.0">
    <dict>
        ....
        <key>CFBundleName</key>
        <string>Something else</string>
        ....
    </dict>
</plist>
```

build and publish the app to `TestApp.app` and then from Finder run the app:

<img width="226" alt="appname" src="https://user-images.githubusercontent.com/2393265/169328986-fc0da34a-aca2-4b5a-a36e-0fea6321f945.png">

You could omit `CFBundleName` from the `Info.plist` in the app bundle, however if you have `CFBundleDisplayName` specified then macOS will use that value as the title for the app.

When you specify both `CFBundleName` and `CFBundleDisplayName` then `CFBundleName` takes precedence.

The Apple Developer docs specify that [`CFBundleName`](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/TP40009249-109585):
> may be displayed to users in situations such as the absence of a value for [CFBundleDisplayName](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-110725) 

Whereas [`CFBundleDisplayName`](https://developer.apple.com/library/archive/documentation/General/Reference/InfoPlistKeyReference/Articles/CoreFoundationKeys.html#//apple_ref/doc/uid/20001431-110725) documentation states:
> specifies the display name of the bundle, visible to users and used by Siri.
Which means that this is a key you should most probably _always_ include in your app bundle.

So to reduce the amount of head-scratching and debugging for other people, my recommendation would be to:

1. Always set `CFBundleDisplayName` in `Info.plist`
2. Set `CFBundleName` to the same value
3. Ensure that `Name` in `App.xaml` is set to the same value

Depending on the build configuration it seems like a smart idea to generate the contents of `Info.plist` at publish time based on the `Name` in `App.xaml`.

I'll create a PR on the Avalonia documentation site to make sure this appears there too.

**Note:** I tried to actually call `AvaloniaNativePlatform.SetupApplicationName()` somewhere after the app starts but whatever you provide there is ignored by macOS.

**Note 2:** This all of course doesn't show up when you launch the app from Rider or the output directory...

**Comment by jeffdoolittle** [comment]
@kekekeks 

Referring to this comment: https://github.com/AvaloniaUI/Avalonia/issues/3541#issuecomment-583905588

Where do you define the command since in this case the binding is happening in App.xaml?

**Comment by timunie** [comment]
Different possibility I think. If it was me I would have an AppViewModel which has a static instance and holds all of my Apps Commands. In the binding you can set that instance as source via x:static

**Comment by kekekeks** [comment]
@jeffdoolittle `App` class can have it's own `DataContext` You are supposed to initialize it in `OnFrameworkInitializationCompleted`

**Comment by timunie** [comment]
Converted this into Q&A as this is neither a bug nor a feature request. If you disagree please let us know.
